class net.sf.jsqlparser.statement.select.SetOperationList cannot be cast to class net.sf.jsqlparser.statement.select.PlainSelect (net.sf.jsqlparser.statement.select.SetOperationList and net.sf.jsqlparser.statement.select.PlainSelect

### 当前使用版本(必填,否则不予处理)
3.5.2


### 该问题是如何引起的？(确定最新版也有问题再提!!!)
使用多租户插件时，执行insert语句会出现一下类型转换异常
class net.sf.jsqlparser.statement.select.SetOperationList cannot be cast to class net.sf.jsqlparser.statement.select.PlainSelect (net.sf.jsqlparser.statement.select.SetOperationList and net.sf.jsqlparser.statement.select.PlainSelect

问题出在TenantLineInnerInterceptor中的一下方法的PlainSelect plainSelect = (PlainSelect)selectBody;行，
protected void processInsertSelect(SelectBody selectBody) {
        PlainSelect plainSelect = (PlainSelect)selectBody;
        FromItem fromItem = plainSelect.getFromItem();
        if (fromItem instanceof Table) {
            this.processPlainSelect(plainSelect);
            this.appendSelectItem(plainSelect.getSelectItems());
        } else if (fromItem instanceof SubSelect) {
            SubSelect subSelect = (SubSelect)fromItem;
            this.appendSelectItem(plainSelect.getSelectItems());
            this.processInsertSelect(subSelect.getSelectBody());
        }

    }
selectBody是SetOperationList 类型



### 重现步骤(如果有就写完整)



### 报错信息
Caused by: java.lang.ClassCastException: class net.sf.jsqlparser.statement.select.SetOperationList cannot be cast to class net.sf.jsqlparser.statement.select.PlainSelect (net.sf.jsqlparser.statement.select.SetOperationList and net.sf.jsqlparser.statement.select.PlainSelect are in unnamed module of loader 'app')
	at com.baomidou.mybatisplus.extension.plugins.inner.TenantLineInnerInterceptor.processInsertSelect(TenantLineInnerInterceptor.java:204)
	at com.baomidou.mybatisplus.extension.plugins.inner.TenantLineInnerInterceptor.processInsert(TenantLineInnerInterceptor.java:138)
	at com.baomidou.mybatisplus.extension.parser.JsqlParserSupport.processParser(JsqlParserSupport.java:89)
	at com.baomidou.mybatisplus.extension.parser.JsqlParserSupport.parserMulti(JsqlParserSupport.java:69)
	at com.baomidou.mybatisplus.extension.plugins.inner.TenantLineInnerInterceptor.beforePrepare(TenantLineInnerInterceptor.java:78)
	at com.baomidou.mybatisplus.extension.plugins.MybatisPlusInterceptor.intercept(MybatisPlusInterceptor.java:102)
	at org.apache.ibatis.plugin.Plugin.invoke(Plugin.java:62)
	at jdk.proxy2/jdk.proxy2.$Proxy146.prepare(Unknown Source)
	at org.apache.ibatis.executor.SimpleExecutor.prepareStatement(SimpleExecutor.java:87)
	at org.apache.ibatis.executor.SimpleExecutor.doUpdate(SimpleExecutor.java:49)
	at org.apache.ibatis.executor.BaseExecutor.update(BaseExecutor.java:117)
	at org.apache.ibatis.executor.CachingExecutor.update(CachingExecutor.java:76)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.apache.ibatis.plugin.Invocation.proceed(Invocation.java:49)
	at com.baomidou.mybatisplus.extension.plugins.MybatisPlusInterceptor.intercept(MybatisPlusInterceptor.java:106)
	at org.apache.ibatis.plugin.Plugin.invoke(Plugin.java:62)
	at jdk.proxy2/jdk.proxy2.$Proxy145.update(Unknown Source)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.apache.ibatis.plugin.Plugin.invoke(Plugin.java:64)
	at jdk.proxy2/jdk.proxy2.$Proxy145.update(Unknown Source)
	at org.apache.ibatis.session.defaults.DefaultSqlSession.update(DefaultSqlSession.java:194)

--- a/mybatis-plus-extension/src/main/java/com/baomidou/mybatisplus/extension/plugins/inner/TenantLineInnerInterceptor.java
+++ b/mybatis-plus-extension/src/main/java/com/baomidou/mybatisplus/extension/plugins/inner/TenantLineInnerInterceptor.java
@@ -129,11 +129,18 @@ public class TenantLineInnerInterceptor extends BaseMultiTableInnerInterceptor i
             Expression tenantId = tenantLineHandler.getTenantId();
             if (itemsList instanceof MultiExpressionList) {
                 ((MultiExpressionList) itemsList).getExpressionLists().forEach(el -> el.getExpressions().add(tenantId));
-            }else if(itemsList instanceof ExpressionList){//fix github issue 4998
+            }else {
                 List<Expression> expressions = ((ExpressionList) itemsList).getExpressions();
-                expressions.forEach(it-> ((RowConstructor)it).getExprList().addExpressions(tenantId));
-            } else {
-                ((ExpressionList) itemsList).getExpressions().add(tenantId);
+                if(CollectionUtils.isNotEmpty(expressions)){//fix github issue 4998 jsqlparse 4.5 批量insert ItemsList不是MultiExpressionList 了，需要特殊处理
+                    Expression expression = expressions.get(0);
+                    if( expression instanceof RowConstructor){
+                        expressions.forEach(it->((RowConstructor)it).getExprList().getExpressions().add(tenantId));
+                    }else{
+                        expressions.add(tenantId);
+                    }
+                }else{
+                    expressions.add(tenantId);
+                }
             }
         } else {
             throw ExceptionUtils.mpe("Failed to process multiple-table update, please exclude the tableName or statementId");

--- a/mybatis-plus-extension/src/main/java/com/baomidou/mybatisplus/extension/plugins/inner/TenantLineInnerInterceptor.java
+++ b/mybatis-plus-extension/src/main/java/com/baomidou/mybatisplus/extension/plugins/inner/TenantLineInnerInterceptor.java
@@ -131,18 +131,20 @@ public class TenantLineInnerInterceptor extends BaseMultiTableInnerInterceptor i
             }else {
                 List<Expression> expressions = ((ExpressionList) itemsList).getExpressions();
                 if(CollectionUtils.isNotEmpty(expressions)){//fix github issue 4998 jsqlparse 4.5 批量insert ItemsList不是MultiExpressionList 了，需要特殊处理
-                    Expression expression = expressions.get(0);
-                    if( expression instanceof RowConstructor){
-                        expressions.forEach(it->((RowConstructor)it).getExprList().getExpressions().add(tenantId));
-                    }else if(expression instanceof Parenthesis){
-                        //Parenthesis 没办法添加新的value了 重置为jdbcParameter  Parenthesis应该只会有1个参数 固定写死
-                        //支持 insert into entity (id) values (?) 这种style
-                        List expressList = new ArrayList();
-                        expressList.add( new JdbcParameter());
-                        ((ExpressionList) itemsList).withExpressions(expressList).withUsingBrackets(true);
-                        ((ExpressionList) itemsList).getExpressions().add(tenantId);
-                    }else{
-                        expressions.add(tenantId);
+                    int len=expressions.size();
+                    for(int i=0;i<len;i++){
+                        Expression expression=expressions.get(i);
+                        if( expression instanceof RowConstructor){
+                            ((RowConstructor) expression).getExprList().getExpressions().add(tenantId);
+                        }else if (expression instanceof Parenthesis){
+                            RowConstructor rowConstructor=new RowConstructor()
+                                .withExprList(new ExpressionList(((Parenthesis) expression).getExpression(),tenantId));
+                            expressions.set(i,rowConstructor);
+                        }else {
+                            if(len-1==i){ // (?,?) 只有最后一个expre的时候才拼接tenantId
+                                expressions.add(tenantId);
+                            }
+                        }
                     }
                 }else{
                     expressions.add(tenantId);

--- a/mybatis-plus-extension/src/main/java/com/baomidou/mybatisplus/extension/plugins/inner/TenantLineInnerInterceptor.java
+++ b/mybatis-plus-extension/src/main/java/com/baomidou/mybatisplus/extension/plugins/inner/TenantLineInnerInterceptor.java
@@ -20,9 +20,7 @@ import com.baomidou.mybatisplus.core.toolkit.*;
 import com.baomidou.mybatisplus.extension.plugins.handler.TenantLineHandler;
 import com.baomidou.mybatisplus.extension.toolkit.PropertyMapper;
 import lombok.*;
-import net.sf.jsqlparser.expression.Expression;
-import net.sf.jsqlparser.expression.RowConstructor;
-import net.sf.jsqlparser.expression.StringValue;
+import net.sf.jsqlparser.expression.*;
 import net.sf.jsqlparser.expression.operators.relational.EqualsTo;
 import net.sf.jsqlparser.expression.operators.relational.ExpressionList;
 import net.sf.jsqlparser.expression.operators.relational.ItemsList;
@@ -43,6 +41,7 @@ import org.apache.ibatis.session.RowBounds;
 
 import java.sql.Connection;
 import java.sql.SQLException;
+import java.util.ArrayList;
 import java.util.List;
 import java.util.Properties;
 
@@ -135,6 +134,13 @@ public class TenantLineInnerInterceptor extends BaseMultiTableInnerInterceptor i
                     Expression expression = expressions.get(0);
                     if( expression instanceof RowConstructor){
                         expressions.forEach(it->((RowConstructor)it).getExprList().getExpressions().add(tenantId));
+                    }else if(expression instanceof Parenthesis){
+                        //Parenthesis 没办法添加新的value了 重置为jdbcParameter  Parenthesis应该只会有1个参数 固定写死
+                        //支持 insert into entity (id) values (?) 这种style
+                        List expressList = new ArrayList();
+                        expressList.add( new JdbcParameter());
+                        ((ExpressionList) itemsList).withExpressions(expressList).withUsingBrackets(true);
+                        ((ExpressionList) itemsList).getExpressions().add(tenantId);
                     }else{
                         expressions.add(tenantId);
                     }

--- a/mybatis-plus-extension/src/main/java/com/baomidou/mybatisplus/extension/plugins/inner/TenantLineInnerInterceptor.java
+++ b/mybatis-plus-extension/src/main/java/com/baomidou/mybatisplus/extension/plugins/inner/TenantLineInnerInterceptor.java
@@ -120,7 +120,7 @@ public class TenantLineInnerInterceptor extends BaseMultiTableInnerInterceptor i
         }
 
         Select select = insert.getSelect();
-        if (select != null) {
+        if (select != null && (select.getSelectBody() instanceof PlainSelect)) { //fix github issue 4998  修复升级到4.5版本的问题
             this.processInsertSelect(select.getSelectBody(), (String) obj);
         } else if (insert.getItemsList() != null) {
             // fixed github pull/295

--- a/mybatis-plus-extension/src/main/java/com/baomidou/mybatisplus/extension/plugins/inner/TenantLineInnerInterceptor.java
+++ b/mybatis-plus-extension/src/main/java/com/baomidou/mybatisplus/extension/plugins/inner/TenantLineInnerInterceptor.java
@@ -21,6 +21,7 @@ import com.baomidou.mybatisplus.extension.plugins.handler.TenantLineHandler;
 import com.baomidou.mybatisplus.extension.toolkit.PropertyMapper;
 import lombok.*;
 import net.sf.jsqlparser.expression.Expression;
+import net.sf.jsqlparser.expression.RowConstructor;
 import net.sf.jsqlparser.expression.StringValue;
 import net.sf.jsqlparser.expression.operators.relational.EqualsTo;
 import net.sf.jsqlparser.expression.operators.relational.ExpressionList;
@@ -128,6 +129,10 @@ public class TenantLineInnerInterceptor extends BaseMultiTableInnerInterceptor i
             Expression tenantId = tenantLineHandler.getTenantId();
             if (itemsList instanceof MultiExpressionList) {
                 ((MultiExpressionList) itemsList).getExpressionLists().forEach(el -> el.getExpressions().add(tenantId));
+            }else if(itemsList instanceof ExpressionList){//fix github issue 4998
+                List<Expression> expressions = ((ExpressionList) itemsList).getExpressions();
+                expressions.forEach(it-> ((RowConstructor)it).getExprList().addExpressions(tenantId));
+
             } else {
                 ((ExpressionList) itemsList).getExpressions().add(tenantId);
             }

--- a/mybatis-plus-extension/src/main/java/com/baomidou/mybatisplus/extension/plugins/inner/TenantLineInnerInterceptor.java
+++ b/mybatis-plus-extension/src/main/java/com/baomidou/mybatisplus/extension/plugins/inner/TenantLineInnerInterceptor.java
@@ -132,7 +132,6 @@ public class TenantLineInnerInterceptor extends BaseMultiTableInnerInterceptor i
             }else if(itemsList instanceof ExpressionList){//fix github issue 4998
                 List<Expression> expressions = ((ExpressionList) itemsList).getExpressions();
                 expressions.forEach(it-> ((RowConstructor)it).getExprList().addExpressions(tenantId));
-
             } else {
                 ((ExpressionList) itemsList).getExpressions().add(tenantId);
             }

--- a/mybatis-plus-extension/src/test/java/com/baomidou/mybatisplus/extension/plugins/inner/TenantLineInnerInterceptorTest.java
+++ b/mybatis-plus-extension/src/test/java/com/baomidou/mybatisplus/extension/plugins/inner/TenantLineInnerInterceptorTest.java
@@ -37,6 +37,8 @@ class TenantLineInnerInterceptorTest {
 
     @Test
     void insert() {
+        assertSql("insert into entity (id) values (?)",
+            "INSERT INTO entity (id, tenant_id) VALUES (?, 1)");
         // plain
         assertSql("insert into entity (id,name) values (?,?)",
             "INSERT INTO entity (id, name, tenant_id) VALUES (?, ?, 1)");

--- a/mybatis-plus-extension/src/test/java/com/baomidou/mybatisplus/extension/plugins/inner/TenantLineInnerInterceptorTest.java
+++ b/mybatis-plus-extension/src/test/java/com/baomidou/mybatisplus/extension/plugins/inner/TenantLineInnerInterceptorTest.java
@@ -39,6 +39,10 @@ class TenantLineInnerInterceptorTest {
     void insert() {
         assertSql("insert into entity (id) values (?)",
             "INSERT INTO entity (id, tenant_id) VALUES (?, 1)");
+        assertSql("insert into entity (id) values (?),(?) ",
+            "INSERT INTO entity (id, tenant_id) VALUES (?, 1), (?, 1)");
+        assertSql("insert into entity (id) values (?),(?) ",
+            "INSERT INTO entity (id, tenant_id) VALUES (?, 1),(?, 1)");
         // plain
         assertSql("insert into entity (id,name) values (?,?)",
             "INSERT INTO entity (id, name, tenant_id) VALUES (?, ?, 1)");

